### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by phoenigm**

### CODE REVIEW

File: backend/src/main/java/ru/kpfu/itis/challenger/security/details/UserDetailsImpl.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import ru.kpfu.itis.challenger.domain.user.State;
+import ru.kpfu.itis.challenger.domain.user.User;
+
+import java.util.Collection;
+import java.util.Collections;
+
+
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/details/UserDetailsImpl.java]: extra

File: backend/src/main/java/ru/kpfu/itis/challenger/security/filters/TokenAuthenticationFilter.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class TokenAuthenticationFilter extends OncePerRequestFilter {
+    public void attemptAuthentication(HttpServletRequest request)
+                                      throws AuthenticationException {
+        String token = request.getHeader(Constants.TOKEN.toString());
+        if (token == null)
+            token = request.getParameter(Constants.TOKEN.toString());
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/filters/TokenAuthenticationFilter.java]: remove 19-20 lines

File: backend/src/main/java/ru/kpfu/itis/challenger/security/handlers/SavedSuccessRequestAwareAuthenticationSuccessHandler.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ 
+    private RequestCache requestCache = new HttpSessionRequestCache();
+ 
+    @Override
+    public void onAuthenticationSuccess(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      Authentication authentication)
+      throws ServletException, IOException {
+
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/handlers/SavedSuccessRequestAwareAuthenticationSuccessHandler.java]: extra

File: backend/src/main/java/ru/kpfu/itis/challenger/security/handlers/SavedSuccessRequestAwareAuthenticationSuccessHandler.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ 
+    @Override
+    public void onAuthenticationSuccess(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      Authentication authentication)
+      throws ServletException, IOException {
+  
+        SavedRequest savedRequest
+          = requestCache.getRequest(request, response);
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/handlers/SavedSuccessRequestAwareAuthenticationSuccessHandler.java]: refactor spaces

File: backend/src/main/java/ru/kpfu/itis/challenger/security/models/TokenAuthentication.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    public TokenAuthentication(String token, Collection<? extends GrantedAuthority> authorities, boolean isAuthenticated,
+                               UserDetailsImpl principal) {
+        this.token = token;
+        this.authorities = authorities;
+        this.isAuthenticated = isAuthenticated;
+        this.principal = principal;
+    }
+
+
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/models/TokenAuthentication.java]: extra

File: backend/src/main/java/ru/kpfu/itis/challenger/security/models/TokenAuthentication.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    @Override
+    public Object getDetails() {
+        return new Object();
+    }
+
+    @Override
+    public String getName() {
+        if (principal != null)
+            return ((UserDetails) principal).getUsername();
+        else
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/models/TokenAuthentication.java]: add {}

File: backend/src/main/java/ru/kpfu/itis/challenger/security/providers/TokenAuthenticationProvider.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    public boolean supports(Class<?> authentication) {
+        return TokenAuthentication.class.equals(authentication);
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        try {
+            TokenAuthentication readyTokenAuthentication = processAuthentication((TokenAuthentication) authentication);
+            return readyTokenAuthentication;
+        } catch (Exception ex) {
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/security/providers/TokenAuthenticationProvider.java]: refactor

File: backend/src/main/java/ru/kpfu/itis/challenger/service/ChallengeService.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+public interface ChallengeService {
+    List<ChallengeDto> getAll();
+
+    List<ChallengeDto> getAllByStatusAndType(Status status, PrivacyType privacyType);
+
+//    List<Challenge> getAllByAuthorUsername(String authorUsername);
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/ChallengeService.java]: remove

File: backend/src/main/java/ru/kpfu/itis/challenger/service/ChallengeServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    @Transactional
+    public ChallengeDto save(NewChallengeForm challengeForm, User user) {
+        Challenge challenge = Challenge.from(challengeForm);
+        challenge.setAuthor(user);
+        challenge.getConditions().forEach(condition -> condition.setChallenge(challenge));
+        challenge.setRegistraionDate(LocalDate.now());
+        if (challengeForm.getParentId() != null) {
+            Optional<Challenge> parentChallenge = challengeRepository.findById(challengeForm.getParentId());
+            parentChallenge.ifPresent(challenge::setParent);
+            // а если нет, то хехехеххе
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/ChallengeServiceImpl.java]: throw logical exception

File: backend/src/main/java/ru/kpfu/itis/challenger/service/ChallengeServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        }
+        challenge.setForkedChallenges(new ArrayList<>());
+        return ChallengeDto.from(challengeRepository.save(challenge));
+    }
+
+    @Override
+    @Transactional
+    public ChallengeDto update(UpdateChallengeForm challengeForm, User user, Long challengeId) {
+        Challenge challengeForUpdate = challengeRepository.getOne(challengeId);
+        if (challengeForUpdate.getAuthor().getId().equals(user.getId())) {
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/ChallengeServiceImpl.java]: replace illegal acc exception and logical code

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+import ru.kpfu.itis.challenger.domain.user.User;
+import ru.kpfu.itis.challenger.repository.SubscriptiontRepository;
+import ru.kpfu.itis.challenger.repository.UserRepository;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+@Transactional
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: remove 23

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+public class FriendsServiceImpl implements FriendsService {
+
+    @Autowired
+    SubscriptiontRepository subscriptionsRepository;
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: private

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        Optional<User> follower = userRepository.findById(request.getFollowerId());
+        Optional<User> followed = userRepository.findById(request.getFollowedId());
+        if (follower.isPresent() && followed.isPresent()) {
+            Subscription subscription = new Subscription();
+            subscription.setFollowed(followed.get());
+            subscription.setFollower(follower.get());
+            subscription.setStatus(SubscriptionStatus.NEW);
+            return SubscriptionDto.from(subscriptionsRepository.save(subscription));
+        }
+        return null;
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: throw logical exceptiolkn

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                        .follower(followed.get())
+                        .status(SubscriptionStatus.ACCEPTED)
+                        .build();
+                subscriptionsRepository.save(reverseSubscription);
+                userToReturn = follower.get();
+            }
+            subscription.setStatus(subscriptionForm.getStatus());
+            subscriptionsRepository.save(subscription);
+        } else {
+            //throw smth
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: throw illegal

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        return Optional.ofNullable(userToReturn);
+    }
+
+    @Override
+    @Transactional
+    public FriendsAndRequestsDTO getFriendsAndRequestsByUser(User user) {
+        User user1 = userRepository.findByLogin(user.getLogin()).get();
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: user1-> user

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+        else if(subscription.getStatus().equals(SubscriptionStatus.DECLINED) ||
+                subscription.getStatus().equals(SubscriptionStatus.NEW)){
+            subscription.setStatus(SubscriptionStatus.ACCEPTED);
+            Subscription newReversedSubscription = Subscription.builder()
+                    .follower(user)
+                    .followed(friend)
+                    .status(SubscriptionStatus.ACCEPTED)
+                    .build();
+            subscriptionsRepository.save(newReversedSubscription);
+            friendStatus= FriendStatus.FRIEND;
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: space after variable

File: backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @Override
+    @Transactional
+    public List<UserDto> getFriendsOfUserByUserId(Long userId) {
+        Optional<User> user = userRepository.findById(userId);
+        return user.map(value -> value.getFollowersSubscriptions().stream().
+                filter(subscription -> subscription.getStatus().equals(SubscriptionStatus.ACCEPTED)).
+                map(subscription -> UserDto.from(subscription.getFollower())).collect(Collectors.toList())).orElse(null);
+//        oreslsethrow smth;
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/FriendsServiceImpl.java]: throw userNotFound

File: backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+    @Autowired
+    private ChallengeRepository challengeRepository;
+    @Autowired
+    private Cloudinary cloudinary;
+
+    @Override
+    public UserDetails loadUserByUsername(String login) throws UsernameNotFoundException {
+        //findByLogin потому что нужно думать как правильно реализовывать findByLoginOrEmail
+        //пока что вроде как в запросе идет условие email=? or username=?
+        // и если у кого-то совпадет email и username, то хз что выдаст
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java]: remove comments

File: backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+
+    public Optional<User> loadUserByLogin(String login) {
+        return userRepository.findByLogin(login);
+    }
+
+    public User register(RegistrationForm form) {
+        if (userRepository.findByLoginOrEmail(form.getLogin(), form.getEmail()).isPresent()) {
+            throw new UserAlreadyExistsException();
+        }
+        // todo: fix user building
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java]: remove

File: backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+        User user = userRepository.findById(id).get();
+        User currentUser = userRepository.findById(authenticationUser.getId()).get();
+        ProfileDto profileDto = getProfileDto(user);
+        profileDto.setFriendStatus(getFrienshipStatus(user, currentUser));
+        return profileDto;
+    }
+
+    private FriendStatus getFrienshipStatus(User user, User currentUser) {
+        List<Subscription> followersSubscriptions = currentUser.getFollowersSubscriptions();
+        List<Subscription> followingSubscriptions = currentUser.getFollowingSubscriptions();
```
Comment (phoenigm) [inline backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java]: ctrl+alt+l

Comment (phoenigm) [reply backend/src/main/java/ru/kpfu/itis/challenger/service/UserService.java]: getStatuses into variable

File: backend/src/main/resources/messages.properties [hunk 21]
```diff
@@ -0,0 +1,2 @@
+registration.username.size = Username should be in 6 to 30
```
Comment (phoenigm) [inline backend/src/main/resources/messages.properties]: compare values with backend

File: frontend/pom.xml [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+					<id>npm run test:unit</id>
+					<goals>
+						<goal>npm</goal>
+					</goals>
+					<phase>test</phase>
+					<configuration>
+						<arguments>run test:unit</arguments>
+					</configuration>
+					</execution>
+					<!-- Run End2End tests with Nightwatch -->
```
Comment (phoenigm) [inline frontend/pom.xml]: delete comments

File: frontend/src/App.vue [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+
+    export default {
+        name: 'app',
+        data() {
+            return {
+                right: null
+            }
+        },
+        created() {
+            document.title = "Look Ma!";
```
Comment (phoenigm) [inline frontend/src/App.vue]: rename to chlanger

File: frontend/src/api/http-common.js [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+    console.log(config);
+
+    return config;
+}, function (error) {
+    console.log('req intercept error: ');
+    console.log(error);
+    return Promise.reject(error);
+});
+
+/*const interceptor = AXIOS.interceptors.response.use(function (response) {
```
Comment (phoenigm) [inline frontend/src/api/http-common.js]: dont forget to rewrite @phoenigm

Comment (phoenigm) [reply frontend/src/api/http-common.js]: gl hf

Comment (phoenigm) [reply frontend/src/api/http-common.js]: DELETE THIS

File: frontend/src/components/ChallengeCard.vue [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+    </router-link>
+</template>
+
+<script>
+    export default {
+        name: "ChallengeCard",
+        props: ['challenge', 'author'],
+        data() {
+            return {
+                /* challenge: {
```
Comment (phoenigm) [inline frontend/src/components/ChallengeCard.vue]: remove!

File: frontend/src/components/FriendCard.vue [hunk 26]
```diff
@@ -0,0 +1,25 @@
+<template>
```
Comment (phoenigm) [inline frontend/src/components/FriendCard.vue]: delete file

File: frontend/src/pages/UserFriends.vue [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+    import {AXIOS} from "../api/http-common";
+
+    export default {
+        name: "UserFriends",
+        components: {NavigationBar},
+        data() {
+            return {
+                friends: [],
+                onlineFriends: [
+                    {active: true, name: 'Jason Oner', avatar: 'https://cdn.vuetifyjs.com/images/lists/1.jpg'},
```
Comment (phoenigm) [inline frontend/src/pages/UserFriends.vue]: remove nafig

File: backend/src/main/java/ru/kpfu/itis/challenger/config/AppConfig.java [hunk 28]
```diff
@@ -0,0 +1,37 @@
...
+                "api_key", apiKey,
+                "api_secret", apiSecret));
+
+        SingletonManager manager = new SingletonManager();
+        manager.setCloudinary(cloudinary);
+        manager.init();
+        return cloudinary;
+    }
+
+
```
[the remark this answers is not in the document]
Comment (emilg1101) [reply backend/src/main/java/ru/kpfu/itis/challenger/config/AppConfig.java]: Extra lineeeeeeeeeeeeeeeeeeeeeeeeeeeeee

[the remark this answers is not in the document]
Comment (phoenigm) [reply backend/src/main/java/ru/kpfu/itis/challenger/config/AppConfig.java]: take the job

File: backend/src/main/java/ru/kpfu/itis/challenger/controller/ChallengeController.java [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+    @PostMapping(path = "/api/challenge")
+    public ResponseEntity save(@Validated @RequestBody NewChallengeForm newChallengeForm, TokenAuthentication authentication) {
+        ChallengeDto challenge = challengeService.save(newChallengeForm, authentication.getUser());
+        return ResponseEntity
+                .created(URI.create("/api/challenge/" + challenge.getId()))
+                .body(challenge);
+    }
+
+    @GetMapping(path = "/api/challenges")
+    @Transactional
```
[the remark this answers is not in the document]
Comment (phoenigm) [reply backend/src/main/java/ru/kpfu/itis/challenger/controller/ChallengeController.java]: take the job

**Pull request merged**